Experimental Async Refreshes API

oneiros opened the pull request:
Internal discussion leading up to this in MAS-429

We have different current and future use-cases where an API request can only deliver partial results and a background job is either triggered or already running that might (!) lead to there being more results later.

Examples of this are:

* Home timeline regeneration
* Fetch all replies
* Discovery FASP searches

We thought a lot about how to add hints for clients without breaking existing APIs. This is a sketch how this could look. In this solution, we add a new, custom HTTP header (`Mastodon-Async-Refresh`) that communicates that there is something running asynchronously that might warrant checking for new results later. A new, experimental API is added to query for the status of such "async refreshes".

~~I am mostly happy with this, except for the name: I think "BackgroundJob" makes it too easy to confuse this with sidekiq jobs or similar. This is *not* trying to re-implement sidekiq's job management and is *not* for every kind of background job.~~ (We renamed this to `AsyncRefresh`).

In this PR I also implemented this for home timeline regeneration, even though it is not strictly needed there (yet). Home timeline regeneration is a simple case, because we currently wait until it is fully finished. There is no number of new results to keep track of. As such there is still a bit missing that the other use-cases will need: Querying the endpoint again while sending along the original async refresh id.

I would have loved to implement a second use-case to include this missing bit as well, but also because I think some of this code can be made more generic, the Redis handling can be encapsulated better and so on. But I would prefer to refactor this according to actual use-cases instead of taking guesses. Sadly there was no time for that, so this will have to wait a bit.

oneiros committed:
Add experimental API to query background jobs

Background job info is kept in redis. This is *NOT* the
same as querying for sidekiq jobs, though, but a bespoke
mechanism to get information about long-running tasks that
may lead to more results being available later.

oneiros committed:
Background job tracking for timeline regeneration

Adds an HTTP header so that clients can query for the status
of the timeline regeneration task.

oneiros committed:
Rename BackgroundJob to AsyncRefresh

To make it crystal clear that this has nothing to do with
(generic, possibly sidekiq based) background jobs.

daprice commented:
Asking as a client dev, has the websocket streaming API been considered as an option for clients to get the results? If I’m understanding this right, there will be some cases where it is better to have the client wait a while and make another request, but in other cases IMO it might make more sense to have the server push the results to the client as soon as it can.

renchap commented:
We need to have an API that works without streaming, because you can not guarantee that your streaming connection is stable and that you will receive all messages there.

It might be useful to add some messages on streaming related to this API, for example sending something when the background job has ended so the client can know it asap (if its using streaming).

Sending new results on streaming will be much harder and might be a bit less useful, as the results are often ordered, so if you get them one by one, you either cant keep the order, or have a sub-par UX where the pages keeps updating until they are all here. This is why we focused on letting the client decide when it reloads the results.

oneiros commented:
Just to add to that: We want this to be usable by all clients on all platforms and there are some that do not use the streaming API at all as far as I know.

daprice commented:
@renchap I agree with that.

Just to be clear, I wasn’t suggesting that streaming should ever be the only way to get anything, just to have it available as an option. I think right now there are already a few parts of the API that weren’t designed with streaming in mind and have to be wastefully polled by the client even when there is a stable streaming connection, and I’m hoping this won’t be another one of those. Having it at least send a streaming message when the job completes makes sense.

oneiros committed:
Improve async refresh management

Try to encapsulate and hide away as much implementation
details as possible.

oneiros committed:
Only add header if async refresh is running

oneiros commented:
> Or alternately, does that violate the "without breaking existing APIs" goal?

Yes, that is exactly it. I am not sure if anyone actually tests for a `200` response specifically in their apps, but they absolutely could and it would not be wrong given the current API docs. So sadly I think we need to consider sending `202` (which I agree would absolutely be better!) a breaking change.

mjankowski commented:
> I think we need to consider sending 202 (which I agree would absolutely be better!) a breaking change.

Got it, thanks for confirming.

I assume you talked about all of this internally, but my next query would be ... given that supporting this new behavior will require client side changes as well to interact with the async refresh results (ie, no one will get a "for free" improved experience), how do the tradeoffs look on bumping the API version (in this home timeline case for example, to `api/v2/timelines/home`) and going the sort-of-standard for this sort of thing http 202 route -vs- optimizing for fewer app changes to expand the existing endpoints with the custom header?

oneiros committed:
Load data from redis once on initialization

And add an AR-style `#reload` method. This should make things
more predictable.

ClearlyClaire reviewed (commented):
That looks good overall! Left a few inline comments.

oneiros committed:
Pipeline redis calls

To save on back and forth communication with redis.

oneiros commented on app/controllers/concerns/async_refreshes_concern.rb:
```diff
@@ -0,0 +1,11 @@
+  def add_async_refresh_header(async_refresh, retry_seconds: 3)
+    return unless async_refresh.running?
+
+    response.headers['Mastodon-Async-Refresh'] = "id=\"#{async_refresh.id}\", retry=#{retry_seconds}"
```
I am not sure what you are referring to, but I deliberately modeled this as a "Dictionary" as per RFC8941 (https://www.rfc-editor.org/rfc/rfc8941.html#name-dictionaries).

Having spent a lot of time in different RFCs over the past months, I am under the impression that this is the "modern" way to structure headers.

oneiros replied:
I certainly hope not.

ClearlyClaire replied:
Then I suppose this is fine!
